README.md
[![npm version](https://badge.fury.io/js/npm-launcher.svg)](https://badge.fury.io/js/npm-launcher)

# npm-launcher

(Former `npm-package-user-scripts-gui`.)

GUI interface to `npm-package-user-scripts-list` command.

The idea of this tool came to my mind when I tried to introduce my son to the
development of node projects and manage them using npm scripts from the
console.

This script scans current project (in running path) `package.json` commands (see
[npm-package-user-scripts-list](https://github.com/lilliputten/npm-package-user-scripts-list)
for parsing details) and shows minimalistic window for interactive running your
service tasks.

Take in mind that the content of the log tab is updated only after the command
is completed. All logs are doubled on stdout.

It may be preferable to install the package globally because each local
installation recompiles its own `libui` library instance.

## Requirements

See [LibUI-Node
Requirements](https://github.com/parro-it/libui-node#prerequisites) -- used
interface engine (required some build tools for compiling sources).

## Installation

For current project:

```shell
$ npm i -S npm-launcher
```
Global:
```shell
$ npm i -g npm-launcher
```

## Usage

In js code:

```js
const ScriptCommandsGUI = require('./');

const commandsWindow = new ScriptCommandsGUI({

  // See options reference below...

  /** Datetime format */
  dateformat: 'yyyy.mm.dd, HH:MM:ss',

  /** Minimal window width */
  width: 800,

});

commandsWindow.showWindow();
```

Linux/MacOS shell command line:
```shell
$ npm-launcher &
```

Windows shell command line (console window will be opened, minimized and closed after program finish):
```shell
$ start /min npm run -s cmd-gui ^& exit
```

Another convenient way to run the program on Windows is with vbasic script like this:
```vbscript
CreateObject("Wscript.Shell").Run "npm run -s cmd-gui", 0
```

(Thanks to Kees Bakker, see his article [How to start Node.js app windowless in
Windows](https://keestalkstech.com/2016/07/start-nodejs-app-windowless-windows/).)

See also scripts `npm-launch.cmd`, `npm-launch.sh`, `npm-launch.vbs` (included
in package) for global-installed package launch.

## Options

All options are optional. :)

<!-- options begin -->
<!-- generated via `scan-options.sh` at 2018.11.27 01:36:25 -->
- **execOptions**: Exec options (for `child_process.exec`); @see [child_process.exec](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback) (default: `{}`)
- **dateformat**: Format datetime for logging; @see [felixge/node-dateformat](https://github.com/felixge/node-dateformat#usage) (default: `'yyyy.mm.dd HH:MM:ss'`)
- **width**: Minimal window width (default: `600`)
- **height**: Minimal window height (default: `200`)
- **title**: Window title (if can't to generate from `package.json`'s `name` field) (default: `'Commands'`)
- **commands**: Commands list (default: `{ test: { title: 'Test' } }`)
- **buttonSpaces**: Spaces for adding before and after button text (default: `' '`)
<!-- options end -->

## Screenshots

Initial screen:

![Initial screen](screenshots/01-initial-screen.png "Initial screen")

Running command:

![Running command](screenshots/02-lint-running.png "Running command")

Successfully finished command:

![Successfully finished command](screenshots/03-lint-done.png "Successfully finished command")

Log tab:

![Log tab](screenshots/04-log-screen.png "Log tab")

Failed command:

![Failed command](screenshots/05-lint-error.png "Failed command")

Show error in the log tab:

![Show error in the log tab](screenshots/06-lint-error-log.png "Show error in the log tab")

## See also:

- [TODO](TODO.md)
- [CHANGELOG](CHANGELOG.md)
- [lilliputten/npm-package-user-scripts-list: List all available npm script commands](https://github.com/lilliputten/npm-package-user-scripts-list)
- [parro-it/libui-node: Node bindings for libui, an awesome native UI library for Unix, OSX and Windows](https://github.com/parro-it/libui-node)

<!--
@version 2018.11.27, 20:24
-->


CHANGELOG.md
# Changelog

- 2018.11.27, 19:54 -- Add links for TODO/CHANGELOG in README.
- 2018.11.27, 19:54 -- Ability to open program window without console. Some hack on Windows -- see in README under 'Usage' section (Kess Bakker's solutuion).
- 2018.11.27, 00:52 -- Flex layout: `Last command` occupy minimal space
- 2018.11.26, 08:56 -- Added available options section in README
- 2018.11.26, 08:45 -- Added last command status field
- 2018.11.26, 08:45 -- Added tabs with 'Commands' & 'Log' section. Log tab contains text field with log and 'Clear' and 'Copy to clipboard' buttons
- 2018.11.26, 09:04 -- CLI interface
- 2018.11.26, 08:47 -- Basic functional: Command buttons, executing actions, logging


test.js
/* eslint-env es6, node, commonjs */
/* eslint-disable no-console */

const ScriptCommandsGUI = require('./');

const commandsWindow = new ScriptCommandsGUI();

// Show window
commandsWindow.showWindow();

// console.log(commandsWindow);
// debugger;


scan-options.sh
#!/bin/sh
# @since 2018.11.26, 22:45
# @version 2018.11.27, 00:26
# @description list options

DATE=`date "+%Y.%m.%d %H:%M:%S"`

TEMPFILE=".temp"
INSERTTO="README.md"

cat "index.js" | sed " \
  :a;N;\$!ba;s/ *\n/<BR>/g; \
  s/^.*\/\*{{{ Propeties\.\.\. \*\/ *<BR>//;
  s/ *\/\* \.\.\.Properties }}}\*\/.*$//;
  s/ *\/\*\* */\n/g;
  s/[\ \t]\+/ /g; \
  s/^\( *<BR> *\)\+//; \
  s/\( *<BR> *\)\+$//; \
  " > $TEMPFILE
LIST=`cat $TEMPFILE | sed 's/[\ \t\n]/<_>/g'`

echo "<!-- generated via \`scan-options.sh\` at $DATE -->" > $TEMPFILE
for S in $LIST; do
  echo "$S" | sed "\
    s/<_>/ /g; \
    s/\(<BR> \+\* \+\|<BR>\) */ /g; \
    s/^ *\(.*\) \+\*\/ \+\(\w\+\): \+\(.*\), *$/- **\2**: \1 (default: \`\3\`)/g; \
    " >> $TEMPFILE
done
echo "" >> $TEMPFILE

RESULT=`cat $TEMPFILE`

PART1=`cat $INSERTTO | tr '\n' '\f' | sed "s/^\(.*<!-- options begin -->\).*$/\1/" | tr '\f' '\n'`
PART2=`cat $INSERTTO | tr '\n' '\f' | sed "s/^.*\(<!-- options end -->.*\)$/\1/" | tr '\f' '\n'`

echo "$PART1" > $INSERTTO
echo "$RESULT" >> $INSERTTO
echo "$PART2" >> $INSERTTO

rm -f $TEMPFILE


index.js
/* eslint-env es6, node, commonjs */
/* eslint-disable no-console */
/**
 * @author lilliputten <lilliputten@yandex.ru>
 * @desc GUI interface to `npm-package-user-scripts-list` command
 * @since 2018.11.26, 03:00
 * @version 2018.11.27, 01:36
 */

const scriptsList = require('npm-package-user-scripts-list');
const path = require('path');
const inherit = require('inherit');
const libui = require('libui-node');
const {exec} = require('child_process');
const dateformat = require('dateformat');
const clipboardy = require('clipboardy');

/** @class ScriptCommandsWindow */
const ScriptCommandsWindow = inherit(/** @lends ScriptCommandsWindow.prototype */{

  /*{{{ Propeties... */

  /** Exec options (for `child_process.exec`);
   * @see [child_process.exec](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback)
   */
  execOptions: {},

  /** Format datetime for logging;
   * @see [felixge/node-dateformat](https://github.com/felixge/node-dateformat#usage)
   */
  dateformat: 'yyyy.mm.dd HH:MM:ss',

  /** Minimal window width */
  width: 600,

  /** Minimal window height */
  height: 200,

  /** Window title (if can't to generate from `package.json`'s `name` field) */
  title: 'Commands',

  /** Commands list */
  commands: { test: { title: 'Test' } },

  /** Spaces for adding before and after button text */
  buttonSpaces: '  ',

  /* ...Properties }}}*/

  // Logging...

  /** getDateTag ** {{{
   */
  getDateTag: function() {
    const now = Date.now();
    const dateTag = dateformat(now, this.dateformat);
    return dateTag;
  },/*}}}*/

  /** writeLog ** {{{
   */
  writeLog: function(type, message) {
    const dateTag = this.getDateTag();
    const text = ( type ? '[' + dateTag + ' ' + type + ']: ' : '' ) + message;
    this.logEntry && this.logEntry.append(text + '\n');
    console.log(text);
  },/*}}}*/
  /** log ** {{{
   * @param {String} message...
   */
  log: function() {
    const message = Array.from(arguments).map(s => String(s)).join(' ');
    this.writeLog('LOG', message);
  },/*}}}*/
  /** error ** {{{
   * @param {String} message...
   */
  error: function() {
    const message = Array.from(arguments).map(s => String(s)).join(' ');
    this.writeLog('ERROR', message);
  },/*}}}*/
  /** clearLog ** {{{
   */
  clearLog: function(){
    this.logEntry && this.logEntry.setText('');
  },/*}}}*/
  /** copyLog ** {{{
   */
  copyLog: function(){
    if (this.logEntry) {
      var text = this.logEntry.getText();
      clipboardy.write(text);
    }
  },/*}}}*/
  /** logDelim ** {{{
   */
  logDelim: function(){
    this.writeLog('', '--------------------------------------------------------------------------------');
  },/*}}}*/

  // Actions...

  /** setActionButtonText ** {{{
   * @param {String} id
   * @param {String} [title]
   */
  setActionButtonText: function (id, title) {
    const action = this.commands[id] || {};
    title = title || action.title;
    if (action.button && action.button.getText() !== title) {
      action.button.setText(this.buttonSpaces + title + this.buttonSpaces);
    }
  },/*}}}*/

  /** actionShowResults ** {{{
   * @param {Object} props
   * @param {String} props.id
   * @param {String} props.status
   * @param {String} [props.title]
   * @param {String} [props.elapsedTime]
   */
  actionShowResults: function(props) {
    const {id} = props;
    const action = this.commands[id] || {};
    if (!props.title || !props.status) {
      props = Object.assign({ title: action.title, status: 'OK' }, props);
    }
    Object.keys(props).map((id) => {
      this.showResultsItemInfo(id, props[id]);
    });
  },/*}}}*/

  /** actionFinished ** {{{
   * @param {String} id
   * @param {String} [title]
   */
  actionFinished: function(id, title) {
    const action = this.commands[id] || {};
    action.runningNow = false;
    this.setActionButtonText(id, title);
    action.button && action.button.setEnabled(true);
  },/*}}}*/

  /** actionStart ** {{{
   * @param {String} id
   */
  actionStart: function(id) {

    const action = this.commands[id];
    const startTime = Date.now();

    const execOptions = Object.assign({}, this.execOptions, {
      // windowsHide: !action.showConsole,
    });

    this.log('Running:', id, '...');
    action.child = exec('npm run -s ' + id + ' 2>&1', execOptions, (error, stdout, stderr) => {
      const endTime = Date.now();
      const elapsedTime = String(endTime - startTime) + ' ms';
      let status = 'SUCCESS';
      if (stdout) {
        this.log('Finished in ' + elapsedTime + '\n' + String(stdout));
      }
      if (error || stderr) {
        this.error(error || stderr);
        /*DEBUG*/debugger;// eslint-disable-line no-debugger
        status = 'ERROR';
      }
      this.logDelim();
      this.actionShowResults({ id, elapsedTime, status });
      this.actionFinished(id);
    });

    action.runningNow = true;
    this.setActionButtonText(id, action.title + ' (Running)');
    action.button && action.button.setEnabled(false);

  },/*}}}*/

  /** buttonAction ** {{{
   * @param {String} id
   */
  buttonAction: function(id) {

    // const action = this.commands[id];

    // if (action.runningNow) {
    //   // TODO: Terminate process
    //   // action.runningNow = false;
    //   title += ' (Canceled)';
    // }

    this.actionStart(id);

  },/*}}}*/

  // Window construction...

  /** placeButtons ** {{{
   * @param {UiVerticalBox} box
   */
  placeButtons: function(box) {

    const group = new libui.UiGroup(' Commands list ');
    group.margined = true;

    const buttons = new libui.UiVerticalBox();

    const scriptKeys = Object.keys(this.commands);
    scriptKeys.map((id) => {
      const action = this.commands[id];
      const title = action.title;
      const buttonTitle = this.buttonSpaces + title + this.buttonSpaces;
      const button = new libui.UiButton(buttonTitle);
      action.button = button;
      button.onClicked(() => this.buttonAction(id));
      buttons.append(button, false);
    });

    group.setChild(buttons);
    box.append(group, true);

  },/*}}}*/
  /** showResultsItemInfo ** {{{
   * @param {String} id
   * @param {String} [value]
   */
  showResultsItemInfo: function(id, value='') {
    if (this.resultElems && this.resultElems[id]) {
      this.resultElems[id].setText(value);
    }
  },/*}}}*/
  /** addResultsItem ** {{{
   * @param {UiContainer} container
   * @param {String} id
   * @param {String} title
   * @param {String} [value]
   */
  addResultsItem: function(container, id, title, value='') {

    const hBox = new libui.UiHorizontalBox();

    const eLabel = new libui.UiLabel(title + ': ');
    hBox.append(eLabel, false);

    const eItem = new libui.UiLabel(value);
    hBox.append(eItem, false);

    container.append(hBox, false);

    // Store element...
    ( this.resultElems || (this.resultElems = {}) )[id] = eItem;

  },/*}}}*/
  /** placeResults ** {{{
   * @param {UiVerticalBox} box
   */
  placeResults: function(box) {

    const group = new libui.UiGroup(' Last command ');
    group.margined = true;

    const container = new libui.UiVerticalBox();

    this.addResultsItem(container, 'status', 'Status');
    this.addResultsItem(container, 'elapsedTime', 'Elapsed time');
    this.addResultsItem(container, 'id', 'Command');
    this.addResultsItem(container, 'title', 'Name');

    group.setChild(container);
    box.append(group, false);

  },/*}}}*/

  /** placeLogBox ** {{{
   * @param {UiVerticalBox} box
   */
  placeLogBox: function(box){

    const logEntry = new libui.UiMultilineEntry();
    box.append(logEntry, true);

    const hBox = new libui.UiHorizontalBox();

    const copyLogButton = new libui.UiButton(' Copy to clipboard ');
    this.copyLog && copyLogButton.onClicked(() => this.copyLog());
    hBox.append(copyLogButton, false);

    const clearLogButton = new libui.UiButton(' Clear all ');
    this.clearLog && clearLogButton.onClicked(() => this.clearLog());
    hBox.append(clearLogButton, false);

    box.append(hBox, false);

    // Save entry for logging functions
    this.logEntry = logEntry;

  },/*}}}*/

  // External interface...

  /** __constructor ** {{{
   * @param {Object} props
   * @param {Number} [props.width] - Window width (0=auto)
   * @param {Number} [props.height] - Window width (0=auto)
   * @param {String} [props.title] - Window title
   * @param {Object} [props.execOptions] - Options for child_process.exec @see [child_process.exec](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback)
   */
  __constructor: function(props = {}) {

    // Evaluate window title from project name (later m.b. overwritten from `props`)
    const pkgFilename = path.join(process.cwd(), 'package.json');
    const pkgData = require(pkgFilename) || {};
    if (pkgData.name) {
      this.prjName = pkgData.name;
      this.title = 'Project [' + this.prjName + '] commands';
    }

    // Load commands list
    this.commands = scriptsList.getScripts();

    // Set all passed params...
    Object.keys(props).map((param) => {
      if (this[param] && props[param]) {
        this[param] = props[param];
      }
    });

  },/*}}}*/

  /** showWindow ** {{{ Show window
   */
  showWindow: function showWindow() {

    const win = new libui.UiWindow(this.title, this.width, this.height, false);
    win.margined = 1;

    win.onClosing(() => {
      win.close();
      libui.stopLoop();
      this.log('EXIT');
    });

    const box = new libui.UiVerticalBox();

    this.placeButtons(box);
    this.placeResults(box);

    const tabs = new libui.UiTab();

    tabs.append('Commands', box);
    const logBox = new libui.UiVerticalBox();
    this.placeLogBox(logBox);
    tabs.append('Log', logBox); // new libui.UiMultilineEntry());

    tabs.setMargined(0, true);
    tabs.setMargined(1, true);

    win.setChild(tabs);
    // win.setChild(box);

    if (this.prjName) {
      this.log('Project:', this.prjName);
    }
    this.log('START');
    this.logDelim();

    win.show();
    libui.startLoop();

  },/*}}}*/

});

module.exports = ScriptCommandsWindow;
